File: backend/src/Tmap.Api/Features/Tasks/TaskDtos.cs
using Tmap.Api.Common;
using Tmap.Api.Features.Subtasks;
using Tmap.Api.Infrastructure.Entities;
using TaskStatus = Tmap.Api.Infrastructure.Entities.TaskStatus;

namespace Tmap.Api.Features.Tasks;

public sealed record CreateTaskRequest(
    Guid? Id,
    string Title,
    string? Notes,
    Guid? ProjectId,
    List<string>? Labels,
    string? Source,
    TaskStatus? Status,
    DateOnly? PlannedDate,
    DateTimeOffset? ScheduledStart,
    DateTimeOffset? ScheduledEnd,
    int? DurationMinutes,
    int? Priority,
    int? ReminderMinutes,
    string? Rank,
    DateOnly? DueDate);

public sealed record UpdateTaskRequest(
    string? Title,
    string? Notes,
    // CLEARABLE: Optional<T?> distinguishes "absent" (leave unchanged) from explicit null (clear),
    // so unschedule / move-to-backlog / rollover persist instead of silently no-op'ing.
    Optional<Guid?> ProjectId,
    List<string>? Labels,
    string? Source,
    TaskStatus? Status,
    Optional<DateOnly?> PlannedDate,
    Optional<DateTimeOffset?> ScheduledStart,
    Optional<DateTimeOffset?> ScheduledEnd,
    int? DurationMinutes,
    int? ActualTimeMinutes,
    Optional<int?> Priority,
    Optional<int?> ReminderMinutes,
    string? Rank,
    DateOnly? DueDate,
    DateTimeOffset? CompletedAt);

public sealed record ReorderItem(Guid Id, string Rank);

public sealed record TaskResponse(
    Guid Id,
    string Title,
    string Notes,
    Guid? ProjectId,
    List<string> Labels,
    string Source,
    TaskStatus Status,
    DateOnly? PlannedDate,
    DateTimeOffset? ScheduledStart,
    DateTimeOffset? ScheduledEnd,
    int DurationMinutes,
    int ActualTimeMinutes,
    int? Priority,
    int? ReminderMinutes,
    string Rank,
    DateOnly? DueDate,
    Guid? RecurrenceRuleId,
    bool IsRecurrenceTemplate,
    bool RecurrenceDetached,
    DateOnly? RecurrenceOriginalDate,
    DateTimeOffset? CompletedAt,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt,
    long ChangeSeq,
    IReadOnlyList<SubtaskResponse> Subtasks);

File: android/app/src/main/java/net/qmindtech/tmap/data/remote/dto/TaskDtos.kt
package net.qmindtech.tmap.data.remote.dto

import kotlinx.serialization.Serializable

@Serializable
data class CreateTaskRequest(
    val id: String? = null,
    val title: String,
    val notes: String? = null,
    val projectId: String? = null,
    val labels: List<String>? = null,
    val source: String? = "android",
    val status: String? = null,
    val plannedDate: String? = null,
    val scheduledStart: String? = null,
    val scheduledEnd: String? = null,
    val durationMinutes: Int? = null,
    val priority: Int? = null,
    val reminderMinutes: Int? = null,
    val rank: String? = null,
    val dueDate: String? = null,
)

@Serializable
data class UpdateTaskRequest(
    val title: String? = null,
    val notes: String? = null,
    val projectId: String? = null,
    val labels: List<String>? = null,
    val source: String? = null,
    val status: String? = null,
    val plannedDate: String? = null,
    val scheduledStart: String? = null,
    val scheduledEnd: String? = null,
    val durationMinutes: Int? = null,
    val priority: Int? = null,
    val reminderMinutes: Int? = null,
    val rank: String? = null,
    val dueDate: String? = null,
)

@Serializable
data class TaskResponse(
    val id: String,
    val title: String,
    val notes: String?,
    val projectId: String?,
    val labels: List<String>? = null,
    val source: String?,
    val status: String,
    val plannedDate: String?,
    val scheduledStart: String?,
    val scheduledEnd: String?,
    val durationMinutes: Int?,
    val actualTimeMinutes: Int,
    val priority: Int?,
    val reminderMinutes: Int?,
    val rank: String?,
    val dueDate: String?,
    val recurrenceRuleId: String?,
    val isRecurrenceTemplate: Boolean,
    val recurrenceDetached: Boolean,
    val recurrenceOriginalDate: String?,
    val completedAt: String?,
    val createdAt: String,
    val updatedAt: String,
    val changeSeq: Long,
    val subtasks: List<SubtaskResponse> = emptyList(),
)
